src/experiment_prototype/experiment_prototype.py

```diff
@@ -951,3 +954,223 @@ def self_check(self):
+                    concurrent = False
+                    concurrent_set = []
+                    for strict_set in strict_slices:
+                        if slice_id in strict_set:
+                            # If the slice id must have the same center freq as another slice
+                            concurrent = True
+                            concurrent_set = strict_set
+
+                    # check if slice is satisfied by current proposed center freq band
+                    # If a freq was in the 50 kHz null in the middle of the band, attempt downshifting
+                    # the center freq
+                    valid = True
+                    if not concurrent:
+                        valid, down_shift = self.compare_freq_to_band(
+                            slice_freq[slice_id], center_band, valid, down_shift
+                        )
+                        if valid:
+                            current_attempt.append(slice_id)
+                        if down_shift:
+                            down_shifting_slice = slice_id
+                    else:
+                        for con_slice in concurrent_set:
+                            valid, down_shift = self.compare_freq_to_band(
+                                slice_freq[con_slice], center_band, valid, down_shift
+                            )
+                        if valid:
+                            for con_slice in concurrent_set:
+                                current_attempt.append(con_slice)
+                        if down_shift:
+                            down_shifting_slice = slice_id
+
+            # Now that all slices have been checked for this lowest freq slice, set center freqs and then
+            # continue the loop if not all slices have been accounted for
+            if down_shifting_slice in current_attempt:
+                # Stop downshifting the freq if down_shifting slice was added to current attempt
+                down_shift = False
+
+            if (
+                down_shift
+                and center_freq <= slice_freq[lowest_freq_slice][0] - tune_bandwidth / 2
+            ):
+                # if downshifting has taken the band below the starting slice freq, stop shifting
+                center_freq += tx_null_band * 2
+                down_shift = False
+                current_attempt = [lowest_freq_slice]
+
+            if down_shift:
+                # if downshifting, shift the center freq back by the null_width
+                center_freq = center_freq - tx_null_band
+            else:
+                # update slices accounted for and assign center freq to slices in current attempt
+                accounted_for = accounted_for.union(set(current_attempt))
+                for slice_id in current_attempt:
+                    if slice_id not in slice_ctr_freq.keys():
+                        slice_ctr_freq[slice_id] = center_freq
+                down_shift = False
```

Comment:
I am finding it tough to wrap my head around what the loop does. I think I was expecting that each time it looped it would add a new set of center frequencies that were allocated, but that's not exactly it, hey? Rather, each iteration may not necessarily add anything, if downshifting is required?

Reply:
Yes. I'm adding more comment to make it clearer here and at the top of the function